libwallet/src/slate_versions/mod.rs

```diff
@@ -80,6 +87,269 @@ impl VersionedSlate {
+			};
+			outputs.push(output);
+		}
+
+		let n_kernels = rdr.read_u16::<BigEndian>()? as usize;
+		let mut kernels: Vec<v2::TxKernelV2> = Vec::with_capacity(n_kernels);
+
+		for _ in 0..n_kernels {
+			let features =
+				v2::KernelFeatures::from_u8(rdr.read_u8()?).ok_or_else(|| ErrorKind::SlateDeser)?;
+			let fee = rdr.read_u64::<BigEndian>()?;
+			let lock_height = rdr.read_u64::<BigEndian>()?;
+
+			let mut excess = [0u8; secp::constants::PEDERSEN_COMMITMENT_SIZE];
+			rdr.read_exact(&mut excess)?;
+			let excess = v2::Commitment(excess);
+
+			let mut excess_sig = [0u8; secp::constants::COMPACT_SIGNATURE_SIZE];
+			rdr.read_exact(&mut excess_sig)?;
+			let excess_sig = v2::Signature::from_raw_data(&excess_sig)?;
+
+			kernels.push(v2::TxKernelV2 {
+				features,
+				fee,
+				lock_height,
+				excess,
+				excess_sig,
+			});
+		}
+
+		let tx = v2::TransactionV2 {
+			offset,
+			body: v2::TransactionBodyV2 {
+				inputs,
+				outputs,
+				kernels,
+			},
+		};
+
+		let amount = rdr.read_u64::<BigEndian>()?;
+		let fee = rdr.read_u64::<BigEndian>()?;
+		let height = rdr.read_u64::<BigEndian>()?;
+		let lock_height = rdr.read_u64::<BigEndian>()?;
+
+		let pdata_len = rdr.read_u16::<BigEndian>()? as usize;
+		let mut participant_data: Vec<v2::ParticipantDataV2> = Vec::with_capacity(pdata_len);
+
+		let s = secp::Secp256k1::new();
+
+		for _ in 0..pdata_len {
+			let id = rdr.read_u8()? as u64;
+
+			let mut public_blind_excess = [0u8; secp::constants::COMPRESSED_PUBLIC_KEY_SIZE];
+			rdr.read_exact(&mut public_blind_excess)?;
+			let public_blind_excess = v2::PublicKey::from_slice(&s, &public_blind_excess)?;
+
+			let mut public_nonce = [0u8; secp::constants::COMPRESSED_PUBLIC_KEY_SIZE];
+			rdr.read_exact(&mut public_nonce)?;
+			let public_nonce = v2::PublicKey::from_slice(&s, &public_nonce)?;
+
+			let part_sig: Option<v2::Signature> = match rdr.read_u8()? {
+				0 => None,
+				_ => {
+					let mut sig = [0u8; 64];
+					rdr.read_exact(&mut sig)?;
+					Some(v2::Signature::from_raw_data(&sig)?)
+				}
+			};
+
+			let message: Option<String> = match rdr.read_u8()? {
+				0 => None,
+				n => {
+					let n = n as usize;
+					let mut msg = [0u8; 255];
+					rdr.read_exact(&mut msg[..n])?;
+					let string =
+						String::from_utf8(msg[..n].to_vec()).map_err(|_| ErrorKind::SlateDeser)?;
+					Some(string)
+				}
+			};
+
+			let message_sig: Option<v2::Signature> = match rdr.read_u8()? {
+				0 => None,
+				_ => {
+					let mut sig = [0u8; 64];
+					rdr.read_exact(&mut sig)?;
+					Some(v2::Signature::from_raw_data(&sig)?)
+				}
+			};
+
+			participant_data.push(v2::ParticipantDataV2 {
+				id,
+				public_blind_excess,
+				public_nonce,
+				part_sig,
+				message,
+				message_sig,
+			});
+		}
+
+		let slate = v2::SlateV2 {
+			version_info,
+			num_participants,
+			id,
+			tx,
+			amount,
+			fee,
+			height,
+			lock_height,
+			participant_data,
+		};
+
+		Ok(VersionedSlate::V2(slate))
+	}
 }
 
```

Comment:
Sorry, I know you asked but I think this should go into the Slate, not the versioned file. Would the assumption be that you would only really want to serialize the latest slate to binary?

Reply:
Why would we only want to serialize the latest? If we choose that route, we have to always be backward-compatible with future changes.

Reply:
Right, in the current setup, it's expected that each new version of the slate will require implementing conversions between previous version->current version and back. So this would require a second implementation to support the binary conversion as well. Perhaps we just want to consider swapping to using one or the other, but not getting into a position where we need to support both.